### PR TITLE
Update Campaigns Execution

### PR DESCRIPTION
**Pull request by rbino**
This PR implements the actual rollout of the Update Campaigns. This makes them functional, with the exception of some corner cases that will be handled in subsequent iterations.

### CODE REVIEW

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 1]
```diff
@@ -0,0 +1,966 @@
...
+
+    test "returns false if campaign has no targets" do
+      update_campaign = update_campaign_fixture()
+
+      assert update_campaign.update_targets == []
+      assert Core.has_idle_targets?(update_campaign.id) == false
+    end
+  end
+
+  test "mark_update_campaign_as_in_progress/1" do
```
Comment (szakhlypa) [inline backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:855]:
```suggestion
  test "mark_update_campaign_as_in_progress!/1" do
```

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 2]
```diff
@@ -0,0 +1,966 @@
...
+    now = DateTime.utc_now()
+
+    assert %UpdateCampaign{status: :in_progress, start_timestamp: ^now} =
+             update_campaign_with_targets_fixture(3)
+             |> Core.mark_update_campaign_as_in_progress!(now)
+  end
+
+  test "mark_update_campaign_as_failed/1" do
```
Comment (szakhlypa) [inline backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:863]:
```suggestion
  test "mark_update_campaign_as_failed!/1" do
```

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 3]
```diff
@@ -0,0 +1,966 @@
...
+    now = DateTime.utc_now()
+
+    assert %UpdateCampaign{status: :finished, outcome: :failure, completion_timestamp: ^now} =
+             update_campaign_with_targets_fixture(3)
+             |> Core.mark_update_campaign_as_failed!(now)
+  end
+
+  test "mark_update_campaign_as_successful/1" do
```
Comment (szakhlypa) [inline backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:871]:
```suggestion
  test "mark_update_campaign_as_successful!/1" do
```

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 4]
```diff
@@ -0,0 +1,966 @@
...
+    test "returns :ok when version is compatible with starting version requirement" do
+      base_image = base_image_fixture(version: "2.3.0", starting_version_requirement: "~> 2.1")
+      target_current_version = Version.parse!("2.2.3")
+      rollout = push_rollout_fixture()
+
+      assert Core.verify_compatibility(target_current_version, base_image, rollout) == :ok
+    end
+  end
+
+  test "mark_target_as_failed/1" do
```
Comment (szakhlypa) [inline backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:407]:
```suggestion
  test "mark_target_as_failed!/1" do
```

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 5]
```diff
@@ -0,0 +1,966 @@
...
+
+    target =
+      target_fixture()
+      |> Core.mark_target_as_failed!(completion_timestamp)
+
+    assert target.status == :failed
+    assert target.completion_timestamp == completion_timestamp
+  end
+
+  test "mark_target_as_successful/1" do
```
Comment (szakhlypa) [inline backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:418]:
```suggestion
  test "mark_target_as_successful!/1" do
```

**Review by szakhlypa — CHANGES_REQUESTED** [review]

**Review by noaccOS — CHANGES_REQUESTED** [review]

File: backend/lib/edgehog/update_campaigns/push_rollout/core.ex [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+  def fetch_target_current_version(target) do
+    device = target.device
+
+    with {:ok, client} <- Devices.appengine_client_from_device(device),
+         {:ok, device_base_image} <- Astarte.fetch_base_image(client, device.device_id) do
+      case Version.parse(device_base_image.version) do
+        {:ok, version} -> {:ok, version}
+        :error -> {:error, :invalid_version}
+      end
+    end
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/core.ex:166]: I find the current version clearer tbh. The `with :error <- something()` works fine with a single match, but imho you lose clarity if success and failure matches are mixed together.

File: backend/lib/edgehog/update_campaigns/push_rollout/core.ex [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+  # it's better to leave everything here and leave refactoring for a second moment when a clear
+  # structure will emerge from the requirements.
+
+  @doc """
+  Gets an UpdateCampaign. Differently from `UpdateCampaigns.fetch_update_campaign/1`, this
+  doesn't preload associations, since they're not needed here.
+  """
+  def get_update_campaign!(update_campaign_id) do
+    Repo.get!(UpdateCampaign, update_campaign_id)
+  end
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/core.ex:45]: The main idea is that `PushRollout.Core` should contain only functions that are used by `PushRollout.Executor`, optimized for this use case. On the other hand, the context functions are the public API exposed to the rest of the application.
Right now I'd prefer to keep the two things separate, because trying to merge them together too early could result in one confusing API instead of two APIs that just do what they need to do. We should at least wait the introduction of another rollout mechanism executor to see more clearly a refactoring path (for example, I expect that at that point this function would be moved in a separate "common" module that all executors can call, since it's not particularly tied to `PushRollou`)

File: backend/lib/edgehog/update_campaigns/push_rollout/core.ex [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+  Returns the BaseImage that belongs to a specific update_campaign_id
+  """
+  def get_update_campaign_base_image!(update_campaign_id) do
+    base_image_query =
+      from uc in UpdateCampaign,
+        join: bi in assoc(uc, :base_image),
+        select: bi
+
+    Repo.get!(base_image_query, update_campaign_id)
+  end
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/core.ex:74]: The code would be more compact (even if I'd use `Ecto.assoc` + `Repo.one!`, because that would better convey the meaning), but this would make this the only function that accepts an `update_campaign` instead of an `update_campaign_id` (other than the ones that actually update the update campaign).
Given the Executor only stores the ID, I'd rather keep the consistency given that we don't actually need all the rest of the struct.

File: backend/lib/edgehog/update_campaigns/push_rollout/executor.ex [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+
+  # Public API
+
+  def start_link(args) do
+    GenStateMachine.start_link(__MODULE__, args)
+  end
+
+  # Callbacks
+
+  @impl true
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/executor.ex:51]: Done

File: backend/test/edgehog/update_campaigns/push_rollout/core_test.exs [hunk 10]
```diff
@@ -0,0 +1,992 @@
...
+  end
+
+  defp update_device_online_for_target(target, online) do
+    Astarte.get_device!(target.device.id)
+    |> Astarte.update_device(%{online: online})
+  end
+
+  defp set_target_retry_count!(target, count) do
+    assert target.retry_count == 0
+    Enum.reduce(0..(count - 1), target, fn _idx, target -> Core.increase_retry_count!(target) end)
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/test/edgehog/update_campaigns/push_rollout/core_test.exs:980]: But counting from 1 kills puppies :dog2:.
JK, done.

File: backend/lib/edgehog/update_campaigns/push_rollout/executor.ex [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+        {:next_state, :rollout, new_data, internal_event(:fetch_next_target)}
+
+      state == :wait_for_campaign_completion and new_data.in_progress_count == 0 ->
+        # We finished updating everything, go to the final state for the finishing touches
+        {:next_state, :campaign_success, new_data}
+
+      true ->
+        # Otherwise, we keep doing what we were doing
+        {:keep_state, new_data}
+    end
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/executor.ex:440]: See above since it's related to that topic

File: backend/lib/edgehog/os_management.ex [hunk 12]
```diff
@@ -163,24 +164,40 @@ defmodule Edgehog.OSManagement do
...
         {:ok, nil}
       end
     end)
     |> Repo.transaction()
     |> case do
       {:ok, %{ota_operation: ota_operation}} ->
-        {:ok, Repo.preload(ota_operation, :device)}
+        ota_operation = Repo.preload(ota_operation, :device)
+
+        PubSub.publish!(:ota_operation_created, ota_operation)
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/os_management.ex:176]: More than failing fast, the point is failing _loud_: `publish!` can fail only in really strange cases, and we want to know it happened

File: backend/lib/edgehog/update_campaigns.ex [hunk 13]
```diff
@@ -452,6 +453,9 @@ defmodule Edgehog.UpdateCampaigns do
       end,
       placeholders: placeholders
     )
+    |> Multi.run(:executor, fn _repo, %{update_campaign: update_campaign} ->
+      ExecutorSupervisor.start_executor(update_campaign)
```
Comment (szakhlypa) [inline backend/lib/edgehog/update_campaigns.ex:457]: Should we start `ExecutorSupervisor` before successful transaction? `PushRollout.Core` may read data from DB (for example in `get_update_campaign!/1`

**Review by szakhlypa — CHANGES_REQUESTED** [review]

File: backend/lib/edgehog/update_campaigns/push_rollout/executor.ex [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+          # OTA Operations to be finished
+          targets_in_progress?(data) ->
+            {:next_state, :wait_for_campaign_completion, data}
+
+          # We're finished
+          true ->
+            {:next_state, :campaign_success, data}
+        end
+    end
+  end
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/push_rollout/executor.ex:204]: I have not moved this to an internal event because while the other instance made sense since it was duplicated code, here I don't think it adds much value

File: backend/lib/edgehog/update_campaigns/executor_supervisor.ex [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+      |> Supervisor.child_spec(id: executor_id)
+
+    DynamicSupervisor.start_child(__MODULE__, child_spec)
+  end
+
+  defp executor_child_spec(%PushRollout{} = _rollout_mechanism, base_args) do
+    # During tests we add `:wait_for_start_execution` to avoid having the executor running
+    # without us being ready to test it
+    args =
+      if Mix.env() == :test do
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/lib/edgehog/update_campaigns/executor_supervisor.ex]: Moved `Mix.env` in the compilation scope so that test arguments are not even compiled in other environments

File: .github/workflows/backend-test.yaml [hunk 16]
```diff
@@ -36,61 +36,88 @@ on:
...
+      run: mix credo
+
+    - name: Check for unused dependencies
+      run: mix do deps.get, deps.unlock --check-unused
 
     # Don't cache PLTs based on mix.lock hash, as Dialyzer can incrementally update even old ones
     # Cache key based on Elixir & Erlang version (also useful when running in matrix)
-    - name: Restore PLT cache
-      uses: actions/cache@v3
+    - name: Cahe Dialyzer's PLT
```
Comment (Pavinati) [inline .github/workflows/backend-test.yaml:91]:
```suggestion
    - name: Cache Dialyzer's PLT
```

File: backend/test/edgehog/update_campaigns/push_rollout/executor_test.exs [hunk 17]
```diff
@@ -0,0 +1,774 @@
...
+        wait_for_state(pid, :wait_for_available_slot)
+      end
+    end
+  end
+
+  describe "PushRollout.Executor marks campaign as successful" do
+    setup do
+      target_count = Enum.random(10..20)
+      # 20 < x <= 70
+      max_errors_percentage = (0.2 + :rand.uniform() * 0.5) * 100
```
Comment (noaccOS) [inline backend/test/edgehog/update_campaigns/push_rollout/executor_test.exs:290]: here too

File: backend/test/edgehog/update_campaigns/push_rollout/executor_test.exs [hunk 18]
```diff
@@ -0,0 +1,774 @@
...
+      |> wait_for_sync!()
+
+      # Expect the Executor to arrive at :wait_for_available_slot
+      wait_for_state(pid, :wait_for_available_slot)
+    end
+
+    test "OTA Requests only to online targets" do
+      target_count = Enum.random(10..20)
+      # We want at least 1 offline target to test that we arrive in :wait_for_target
+      online_count = Enum.random(0..(target_count - 1))
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/test/edgehog/update_campaigns/push_rollout/executor_test.exs:178]: This was actually correct (see the comment above), but I changed it to randomize the offline count (with your formula) and deriving the offline count from `target_count - offline_count` so it's more direct and clearer

**Review by szakhlypa — APPROVED** [review]

**Review by noaccOS — APPROVED** [review]
